src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Windows.cs

```diff
@@ -41,58 +45,115 @@ private static FileStreamStrategy ChooseStrategyCore(SafeFileHandle handle, File
+                Debug.Assert(path != null);
 
-            int fAccess =
-                ((access & FileAccess.Read) == FileAccess.Read ? Interop.Kernel32.GenericOperations.GENERIC_READ : 0) |
-                ((access & FileAccess.Write) == FileAccess.Write ? Interop.Kernel32.GenericOperations.GENERIC_WRITE : 0);
+                uint ntStatus = 0;
+                if (ShouldPreallocate(allocationSize, access, mode))
+                {
+                    GetPathForNtCreateFile(path, out ReadOnlySpan<char> prefixedAbsolutePath, out char[]? rentedArray);
 
-            // Our Inheritable bit was stolen from Windows, but should be set in
-            // the security attributes class.  Don't leave this bit set.
-            share &= ~FileShare.Inheritable;
+                    (ntStatus, IntPtr fileHandle) = Interop.NtDll.CreateFile(prefixedAbsolutePath, mode, access, share, options, allocationSize);
 
-            // Must use a valid Win32 constant here...
-            if (mode == FileMode.Append)
-                mode = FileMode.OpenOrCreate;
+                    if (rentedArray is not null)
+                    {
+                        ArrayPool<char>.Shared.Return(rentedArray);
+                    }
 
-            int flagsAndAttributes = (int)options;
+                    if (ntStatus == 0)
+                    {
+                        return ValidateFileHandle(new SafeFileHandle(fileHandle, ownsHandle: true), path, (options & FileOptions.Asynchronous) != 0);
+                    }
+                    else if (ntStatus == NT_ERROR_STATUS_DISK_FULL)
+                    {
+                        throw new IOException(SR.Format(SR.IO_DiskFull_Path_AllocationSize, path, allocationSize));
+                    }
+                    else if (ntStatus == NT_ERROR_STATUS_FILE_TOO_LARGE)
+                    {
+                        throw new IOException(SR.Format(SR.IO_FileTooLarge_Path_AllocationSize, path, allocationSize));
+                    }
 
-            // For mitigating local elevation of privilege attack through named pipes
-            // make sure we always call CreateFile with SECURITY_ANONYMOUS so that the
-            // named pipe server can't impersonate a high privileged client security context
-            // (note that this is the effective default on CreateFile2)
-            flagsAndAttributes |= (Interop.Kernel32.SecurityOptions.SECURITY_SQOS_PRESENT | Interop.Kernel32.SecurityOptions.SECURITY_ANONYMOUS);
+                    // NtCreateFile has failed for some other reason than a full disk or too large file.
+                    // Instead of implementing the mapping for every NS Status value (there are plenty of them: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55)
+                    // or using RtlNtStatusToDosError & GetExceptionForWin32Error
+                    // the code falls back to CreateFileW that just throws the right exception.
```

Comment:
> Is that expected?

Yes, it is. The main goal here is to throw a new exception for `DISK_FULL` and `FILE_TOO_LARGE`, but the "old" one for other cases (like when file is locked, user don't have access, file does not exist etc)

> Could that re-try cripple perf on devices with small disk size

It's only going to slightly affect the performance of exceptional code paths, which in general are very slow because exception handling is very expensive.

> (or where NtDll.CreateFile could always fail for some reason

this should never be the case as internally `CreateFile` also calls `NtCreateFile`

Reply:
after reading Jans comment I have changed that and we throw immediately